service-proxy.js

```diff
@@ -675,6 +705,36 @@ function ensurePeerConnected(serviceName, peer, reason, now) {
         peer.clearDrain('canceled to ensure peer connection');
     }
 
+    self.peersToConnect[peer.hostPort] = {
+        lastRefresh: now,
+        serviceName: serviceName,
+        reason: reason
+    };
+};
+
+ServiceDispatchHandler.prototype.connectSinglePeer =
+function connectSinglePeer(hostPort, connectInfo) {
+    var self = this;
+
+    if (self.peersToPrune[hostPort]) {
+        return;
+    }
+
+    var serviceName = connectInfo.serviceName;
+    var serviceChannel = self.getServiceChannel(serviceName);
+    if (!serviceChannel) {
```

Comment:
I see we do allocate the peer; so if the peer was de-allocated then we do not connect.

Are there any race conditions here where we remove peers. I dont think there are.